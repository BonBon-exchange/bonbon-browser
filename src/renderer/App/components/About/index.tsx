import { About } from './About';

export * from './About';
export default About;
