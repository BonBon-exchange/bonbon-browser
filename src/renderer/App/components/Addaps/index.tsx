import { Addaps } from './Addaps';

export * from './Addaps';
export default Addaps;
