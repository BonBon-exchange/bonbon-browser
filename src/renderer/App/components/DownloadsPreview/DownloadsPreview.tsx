/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable import/prefer-default-export */
import { Button } from '@mui/material';
import LinearProgress from '@mui/material/LinearProgress';
import { useTranslation } from 'react-i18next';

import { useAppSelector, useAppDispatch } from 'renderer/App/store/hooks';

import {
  DownloadItem,
  DownloadsState,
  clearDownloads,
} from 'renderer/App/store/reducers/Downloads';

import './style.scss';

export const DownloadsPreview: React.FC = () => {
  const { t } = useTranslation();
  const { items }: DownloadsState = useAppSelector((state) => state.downloads);
  const dispatch = useAppDispatch();

  const handleOnClick = (i: DownloadItem) => {
    window.app.tools.showItemInFolder(i.savePath);
  };

  const handleClear = () => {
    dispatch(clearDownloads());
    window.app.download.hideDownloadsPreview();
  };

  return (
    <div id="DownloadsPreview__container">
      <ul>
        {items.map((i) => {
          return (
            <li
              onClick={() => handleOnClick(i)}
              key={`${i.etag}::${i.startTime}`}
            >
              <div className="DownloadsPreview__item-name">{i.filename}</div>
              {i.state !== 'progressing' && i.state !== 'completed' && (
                <div className="DownloadsPreview__item-state">
                  {i.state ? t(i.state) : ''}
                </div>
              )}
              {i.state !== 'completed' && (
                <LinearProgress
                  variant="determinate"
                  value={i.progress * 100}
                  className="DownloadsPreview__item-progress"
                />
              )}
            </li>
          );
        })}
      </ul>
      <div className="DownloadsPreview__clear" onClick={handleClear}>
        <Button variant="contained">Clear</Button>
      </div>
    </div>
  );
};
