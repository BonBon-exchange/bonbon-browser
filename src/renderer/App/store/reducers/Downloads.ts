import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { DownloadState } from 'renderer/TitleBar/components/TopBar/Types';

export type DownloadItem = {
  savePath: string;
  filename: string;
  progress: number;
  etag: string;
  startTime: number;
  state: DownloadState;
};

export type DownloadsState = {
  items: DownloadItem[];
};

export const initialState: DownloadsState = {
  items: [],
};

export const downloadsSlice = createSlice({
  name: 'downloads',
  initialState,
  reducers: {
    setDownloadItem: (state, action: PayloadAction<DownloadItem>) => {
      const itemIndex = state.items.findIndex(
        (i) =>
          i.savePath === action.payload.savePath &&
          i.startTime === action.payload.startTime
      );
      if (itemIndex === -1) state.items.push(action.payload);
      else state.items[itemIndex] = action.payload;

      if (state.items.length > 10) state.items.splice(0, 1);
    },
    clearDownloads: (state) => {
      state.items = [];
    },
  },
});

export const { setDownloadItem, clearDownloads } = downloadsSlice.actions;

export default downloadsSlice.reducer;
