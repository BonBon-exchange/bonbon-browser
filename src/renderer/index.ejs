<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta
      http-equiv="Content-Security-Policy"
    />
    <title>BonBon</title>
  </head>
  <body>
    <div id="root"></div>
  </body>
</html>
