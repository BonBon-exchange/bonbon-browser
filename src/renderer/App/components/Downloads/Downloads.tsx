/* eslint-disable promise/always-return */
/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable import/prefer-default-export */
import { useState, useEffect } from 'react';
import { useTranslation } from 'react-i18next';
import DeleteForeverIcon from '@mui/icons-material/DeleteForever';
import Button from '@mui/material/Button';

import { CloseButton } from 'renderer/App/components/CloseButton';

import { DownloadsProps, DownloadType } from './Types';

import './style.scss';

export const Downloads: React.FC<DownloadsProps> = ({
  handleClose,
}: DownloadsProps) => {
  const { t } = useTranslation();
  const [search, setSearch] = useState<string>('');
  const [items, setItems] = useState<DownloadType[]>([]);
  const [filteredItems, setFilteredItems] = useState<DownloadType[]>([]);

  const handleOpenDownload = (savePath: string) => {
    window.app.tools.showItemInFolder(savePath);
  };

  const handleDeleteDownload = (id: number) => {
    window.app.download
      .removeDownload(id)
      .then(() => {
        const newItems = [...items];
        const index = newItems.findIndex((i) => i.id === id);
        if (index > -1) newItems.splice(index, 1);
        setItems(newItems);
      })
      .catch(console.log);
  };

  const handleClearDownloads = () => {
    window.app.download
      .clearDownloads()
      .then(() => {
        setItems([]);
      })
      .catch(console.log);
  };

  useEffect(() => {
    const filtered = items.filter(
      (i) =>
        (i.filename && i.filename.includes(search)) ||
        (i.savePath && i.savePath.includes(search))
    );
    setFilteredItems(filtered);
  }, [search, items]);

  useEffect(() => {
    window.app.download
      .getAllDownloads()
      .then((val) => {
        setItems(val);
        setFilteredItems(val);
      })
      .catch(console.log);

    window.app.analytics.page('/downloads');
  }, []);

  return (
    <div id="Downloads__container">
      <CloseButton handleClose={handleClose} />
      <div id="Downloads__centered-container">
        <h2>{t('Downloads')}</h2>
        <Button
          variant="contained"
          className="Downloads__clear-button"
          onClick={handleClearDownloads}
        >
          {t('Clear')}
        </Button>
        <input
          type="text"
          className="Downloads__search"
          onChange={(e) => setSearch(e.target.value)}
          placeholder={t('Search')}
        />
        {filteredItems.map((i) => {
          return (
            <div className="Downloads__item" key={i.id}>
              <div
                className="Downloads__item-text"
                onClick={() => handleOpenDownload(i.savePath)}
              >
                <div className="Downloads__item-title">{i.filename}</div>
                <div className="Downloads__item-date">{i.date}</div>
                <div className="Downloads__item-url">{i.savePath}</div>
              </div>
              <div className="Downloads__item-controls">
                <div
                  className="Downloads__item-control"
                  onClick={() => handleDeleteDownload(i.id)}
                >
                  <DeleteForeverIcon />
                </div>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};
