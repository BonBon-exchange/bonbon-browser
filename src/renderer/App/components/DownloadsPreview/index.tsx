import { DownloadsPreview } from './DownloadsPreview';

export * from './DownloadsPreview';
export default DownloadsPreview;
