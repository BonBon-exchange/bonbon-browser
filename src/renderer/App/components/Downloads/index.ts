import { Downloads } from './Downloads';

export * from './Downloads';
export default Downloads;
