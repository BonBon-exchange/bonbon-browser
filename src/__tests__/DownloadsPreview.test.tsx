/* eslint-disable react/jsx-props-no-spreading */
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { Middleware } from '@reduxjs/toolkit';
import renderer, { act } from 'react-test-renderer';

import { DownloadsPreview } from '../renderer/App/components/DownloadsPreview';
import { mockWindow } from './beforeAll';
import { initialState } from '../renderer/App/store/reducers/Downloads';

let tree: any;
let store: any;
const middlewares: Middleware[] = [];

describe('DownloadsPreview', () => {
  beforeAll(() => {
    mockWindow();
    const mockStore = configureStore(middlewares);
    store = mockStore({ downloads: initialState });
  });

  it('should render', () => {
    act(() => {
      tree = renderer.create(
        <Provider store={store}>
          <DownloadsPreview />
        </Provider>
      );
    });

    expect(tree).toMatchSnapshot();
  });
});
