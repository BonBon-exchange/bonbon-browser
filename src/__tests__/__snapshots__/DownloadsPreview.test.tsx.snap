// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadsPreview should render 1`] = `
<div
  id="DownloadsPreview__container"
>
  <ul />
  <div
    className="DownloadsPreview__clear"
    onClick={[Function]}
  >
    <button
      className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Clear
      <span
        className="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
